torch/utils/inference/run.py

```diff
@@ -0,0 +1,569 @@
+
+1. Run single-instance inference on a single node with all CPU nodes.
+
+::
+
+   >>> torchrun --inference --throughput_mode script.py args
+
+2. Run single-instance inference on a single CPU node.
+
+::
+
+   >>> torchrun --inference --node_id 1 script.py args
+
+*** Multi-instance inference ***
+
+1. Multi-instance
+   By default, one instance per node. if you want to set the instance numbers and core per instance,
+   --nprocs and  --ncore_per_instance should be set.
+
+::
+
+   >>> torchrun --inference -- python_script args
+
+   eg: on an Intel(R) Xeon(R) Scalable Processor with 14 instance, 4 cores per instance
+
+::
+
+   >>> torchrun --inference  --nprocs 14 --ncore_per_instance 4 python_script args
+
+2. Run single-instance inference among multiple instances.
+   By default, runs all nprocs. If you want to independently run a single instance among nprocs, specify instance_idx.
+
+   eg: run 0th instance on an Intel(R) Xeon(R) Scalable Processor with 2 instance (i.e., numactl -C 0-27)
+
+::
+
+   >>> torchrun --inference  --nprocs 2 --instance_idx 0 python_script args
+
+   eg: run 1st instance on an Intel(R) Xeon(R) Scalable Processor with 2 instance (i.e., numactl -C 28-55)
+
+::
+
+   >>> torchrun --inference  --nprocs 2 --instance_idx 1 python_script args
+
+   eg: run 0th instance on an Intel(R) Xeon(R) Scalable Processor with 2 instance, 2 cores per instance,
+   first four cores (i.e., numactl -C 0-1)
+
+::
+
+   >>> torchrun --inference  --core_list "0, 1, 2, 3" --nprocs 2 --ncore_per_instance 2 --instance_idx 0 python_script args
+
+3. To look up what optional arguments this module offers:
+
+::
+
+    >>> torchrun --inference --help
+
+*** Memory allocator  ***
+
+"--enable_tcmalloc" and "--enable_jemalloc" can be used to enable different memory allcator.
+
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+import sys
+import platform
+import subprocess
+import os
+from os.path import expanduser
+import re
+import glob
+from argparse import ArgumentParser, REMAINDER
+from argparse import RawTextHelpFormatter
+import logging
+from torch.distributed.elastic.multiprocessing import Std, start_processes
+from typing import List, Dict
+
+format_str = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+logging.basicConfig(level=logging.INFO, format=format_str)
+logger = logging.getLogger(__name__)
+# from torch.distributed.elastic.utils.logging import get_logger
+# logger = get_logger(__name__)
+
+class _CPUinfo():
+    """
+    Get CPU inforamation, such as cores list and NUMA information.
+    """
+    def __init__(self):
+
+        self.cpuinfo = []
+        if platform.system() in ["Windows", "Darwin"]:
+            raise RuntimeError(f"{platform.system()} is not supported!!!")
+        elif platform.system() == "Linux":
+            args = ["lscpu", "--parse=CPU,Core,Socket,Node"]
+            lscpu_info = subprocess.check_output(args, universal_newlines=True).split("\n")
```

Comment:
Lets add a comment here with a sample output of running lscpu, makes it easier to follow the rest of the logic. Something like

```
# Sample output of: `lscpu --parse=CPU,Core,Socket,Node`
# 
# # The following is the parsable format, which can be fed to other
# # programs. Each different item in every column has an unique ID
# # starting from zero.
# # CPU,Core,Socket,Node
# 0,0,0,0
# 1,1,0,0
...
```